File: packages/sandboxd/src/tunnel/telemetry.rs
//! Telemetry stream buffering for the bootstrap tunnel.
//!
//! The sandbox supervisor emits its own diagnostic lines locally, but the
//! gateway can subscribe to a bounded subset of those logs over the dedicated
//! telemetry stream. This module owns the telemetry stream state machine,
//! send-window accounting, and bounded buffering needed to publish those log
//! lines over the existing bootstrap websocket.

use std::fmt::{self, Display};

use serde_json::{Map, Value};

use crate::time::{Clock, format_rfc3339_timestamp};
use crate::tunnel::protocol::{
    BootstrapTelemetryControlMessage, MAX_STREAM_WINDOW_BYTES, PAYLOAD_KIND_RAW_BYTES,
    StreamSendWindow, decode_stream_data_frame, encode_stream_data_frame,
    parse_bootstrap_telemetry_control_message, telemetry_close, telemetry_open,
};

/// Reserved bootstrap-tunnel stream id for sandbox diagnostics.
pub const SANDBOX_TELEMETRY_LOG_STREAM_ID: u32 = 0xffff_fffe;
const TELEMETRY_LOGS_SIGNAL: &str = "logs";
const TELEMETRY_LOGS_FORMAT: &str = "mistle.sandbox-runtime.log.v1";
const RESERVED_LOG_FIELDS: [&str; 3] = ["timestamp", "level", "event"];

/// Supported log severities for sandbox telemetry log lines.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SandboxTelemetryLogLevel {
    Info,
    Warn,
    Error,
}

impl SandboxTelemetryLogLevel {
    fn as_str(self) -> &'static str {
        match self {
            Self::Info => "info",
            Self::Warn => "warn",
            Self::Error => "error",
        }
    }
}

/// Describes why one telemetry relay step failed.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TelemetryRelayError {
    message: String,
}

impl TelemetryRelayError {
    fn new(message: impl Into<String>) -> Self {
        Self {
            message: message.into(),
        }
    }
}

impl Display for TelemetryRelayError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(&self.message)
    }
}

impl std::error::Error for TelemetryRelayError {}

/// One serialized bootstrap-tunnel frame produced by the telemetry relay.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum TelemetryRelayFrame {
    Text(String),
    Binary(Vec<u8>),
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum TelemetryRelayState {
    Disconnected,
    Opening,
    Open,
}

/// Buffers telemetry log lines until the gateway has granted enough send window.
#[derive(Debug)]
pub struct TelemetryRelay {
    state: TelemetryRelayState,
    send_window: StreamSendWindow,
    buffered_lines: Vec<Vec<u8>>,
    buffered_bytes: usize,
}

impl Default for TelemetryRelay {
    fn default() -> Self {
        Self {
            state: TelemetryRelayState::Disconnected,
            send_window: StreamSendWindow::new(0),
            buffered_lines: Vec::new(),
            buffered_bytes: 0,
        }
    }
}

impl TelemetryRelay {
    /// Starts telemetry negotiation on the current tunnel connection.
    pub fn attach_tunnel_connection(
        &mut self,
    ) -> Result<Vec<TelemetryRelayFrame>, TelemetryRelayError> {
        self.state = TelemetryRelayState::Opening;
        self.send_window = StreamSendWindow::new(0);
        self.buffered_lines.clear();
        self.buffered_bytes = 0;
        Ok(vec![TelemetryRelayFrame::Text(telemetry_open(
            SANDBOX_TELEMETRY_LOG_STREAM_ID,
            TELEMETRY_LOGS_SIGNAL,
            TELEMETRY_LOGS_FORMAT,
        ))])
    }

    /// Stops telemetry publication on the current tunnel connection.
    pub fn detach_tunnel_connection(
        &mut self,
    ) -> Result<Vec<TelemetryRelayFrame>, TelemetryRelayError> {
        let mut frames = Vec::new();
        if matches!(
            self.state,
            TelemetryRelayState::Opening | TelemetryRelayState::Open
        ) {
            frames.push(TelemetryRelayFrame::Text(telemetry_close(
                SANDBOX_TELEMETRY_LOG_STREAM_ID,
            )));
        }

        self.state = TelemetryRelayState::Disconnected;
        self.send_window = StreamSendWindow::new(0);
        self.buffered_lines.clear();
        self.buffered_bytes = 0;
        Ok(frames)
    }

    /// Applies one telemetry control response from the gateway.
    pub fn handle_control_message(
        &mut self,
        payload: &str,
    ) -> Result<Option<Vec<TelemetryRelayFrame>>, TelemetryRelayError> {
        let Some(message) = parse_bootstrap_telemetry_control_message(payload)
            .map_err(|error| TelemetryRelayError::new(error.to_string()))?
        else {
            return Ok(None);
        };

        match message {
            BootstrapTelemetryControlMessage::OpenOk(message) => {
                if message.stream_id != SANDBOX_TELEMETRY_LOG_STREAM_ID {
                    return Ok(None);
                }
                self.send_window = StreamSendWindow::new(message.initial_window_bytes);
                self.state = TelemetryRelayState::Open;
                Ok(Some(self.flush()?))
            }
            BootstrapTelemetryControlMessage::Window(message) => {
                if message.stream_id != SANDBOX_TELEMETRY_LOG_STREAM_ID {
                    return Ok(None);
                }
                self.send_window
                    .add(message.bytes)
                    .map_err(|error| TelemetryRelayError::new(error.to_string()))?;
                Ok(Some(self.flush()?))
            }
            BootstrapTelemetryControlMessage::OpenError(message) => {
                if message.stream_id != SANDBOX_TELEMETRY_LOG_STREAM_ID {
                    return Ok(None);
                }
                Err(TelemetryRelayError::new(format!(
                    "gateway rejected telemetry stream: {} ({})",
                    message.message, message.code
                )))
            }
            BootstrapTelemetryControlMessage::Reset(message) => {
                if message.stream_id != SANDBOX_TELEMETRY_LOG_STREAM_ID {
                    return Ok(None);
                }
                Err(TelemetryRelayError::new(format!(
                    "gateway reset telemetry stream: {} ({})",
                    message.message, message.code
                )))
            }
        }
    }

    /// Queues one telemetry log line for transmission.
    pub fn enqueue_log_line(
        &mut self,
        line: &str,
    ) -> Result<Vec<TelemetryRelayFrame>, TelemetryRelayError> {
        if self.buffered_bytes.saturating_add(line.len()) > MAX_STREAM_WINDOW_BYTES {
            return Err(TelemetryRelayError::new(
                "telemetry buffer exceeded the configured capacity",
            ));
        }

        self.buffered_bytes = self.buffered_bytes.saturating_add(line.len());
        self.buffered_lines.push(line.as_bytes().to_vec());
        self.flush()
    }

    /// Serializes one structured telemetry log record and queues it for transmission.
    pub fn enqueue_log_record(
        &mut self,
        clock: &dyn Clock,
        level: SandboxTelemetryLogLevel,
        event: &str,
        extra_fields: &[(&str, Value)],
    ) -> Result<Vec<TelemetryRelayFrame>, TelemetryRelayError> {
        let line = encode_sandbox_telemetry_log_line(clock, level, event, extra_fields)?;
        self.enqueue_log_line(&line)
    }

    fn flush(&mut self) -> Result<Vec<TelemetryRelayFrame>, TelemetryRelayError> {
        if self.state != TelemetryRelayState::Open {
            return Ok(Vec::new());
        }

        let mut frames = Vec::new();
        while let Some(next_line) = self.buffered_lines.first() {
            if !self.send_window.try_consume(next_line.len()) {
                return Ok(frames);
            }

            let line = self.buffered_lines.remove(0);
            self.buffered_bytes = self.buffered_bytes.saturating_sub(line.len());
            let encoded = encode_stream_data_frame(
                SANDBOX_TELEMETRY_LOG_STREAM_ID,
                PAYLOAD_KIND_RAW_BYTES,
                &line,
            )
            .map_err(|error| TelemetryRelayError::new(error.to_string()))?;
            frames.push(TelemetryRelayFrame::Binary(encoded));
        }

        Ok(frames)
    }
}

/// Decodes one telemetry data frame routed back from the tunnel.
pub fn decode_telemetry_data_frame(payload: &[u8]) -> Result<Vec<u8>, TelemetryRelayError> {
    let frame = decode_stream_data_frame(payload)
        .map_err(|error| TelemetryRelayError::new(error.to_string()))?;
    if frame.stream_id != SANDBOX_TELEMETRY_LOG_STREAM_ID {
        return Err(TelemetryRelayError::new(format!(
            "unexpected telemetry streamId {}",
            frame.stream_id
        )));
    }
    if frame.payload_kind != PAYLOAD_KIND_RAW_BYTES {
        return Err(TelemetryRelayError::new(
            "telemetry stream only accepts raw byte payloads",
        ));
    }
    Ok(frame.payload)
}

/// Serializes one newline-delimited `mistle.sandbox-runtime.log.v1` log line.
pub fn encode_sandbox_telemetry_log_line(
    clock: &dyn Clock,
    level: SandboxTelemetryLogLevel,
    event: &str,
    extra_fields: &[(&str, Value)],
) -> Result<String, TelemetryRelayError> {
    if event.trim().is_empty() {
        return Err(TelemetryRelayError::new(
            "sandbox telemetry log event must not be empty",
        ));
    }

    let mut payload = Map::new();
    payload.insert(
        "timestamp".to_string(),
        Value::String(
            format_rfc3339_timestamp(clock.now_system_time())
                .map_err(|error| TelemetryRelayError::new(error.to_string()))?,
        ),
    );
    payload.insert(
        "level".to_string(),
        Value::String(level.as_str().to_string()),
    );
    payload.insert("event".to_string(), Value::String(event.to_string()));

    for (field_name, field_value) in extra_fields {
        if field_name.trim().is_empty() {
            return Err(TelemetryRelayError::new(
                "sandbox telemetry log field name must not be empty",
            ));
        }
        if RESERVED_LOG_FIELDS.contains(field_name) {
            return Err(TelemetryRelayError::new(format!(
                "sandbox telemetry log field '{field_name}' is reserved",
            )));
        }
        if !matches!(
            field_value,
            Value::Null | Value::Bool(_) | Value::Number(_) | Value::String(_)
        ) {
            return Err(TelemetryRelayError::new(format!(
                "sandbox telemetry log field '{field_name}' must be scalar",
            )));
        }

        payload.insert((*field_name).to_string(), field_value.clone());
    }

    let mut line = serde_json::to_string(&Value::Object(payload))
        .map_err(|error| TelemetryRelayError::new(error.to_string()))?;
    line.push('\n');
    Ok(line)
}

File: packages/sandboxd/tunnel/telemetry.go
package tunnel

import (
	"context"
	"encoding/json"
	"fmt"
	"os"

	"github.com/coder/websocket"
	"github.com/mistle/sandboxd/supervision"
	"github.com/mistle/sandboxd/timeutil"
	tunnelprotocol "github.com/mistle/sandboxd/tunnel/protocol"
)

const (
	SandboxTelemetryLogStreamID   = uint32(0xffff_fffe)
	telemetryLogsSignal           = "logs"
	telemetryLogsFormat           = "mistle.sandbox-runtime.log.v1"
	telemetryBufferCapacityBytes  = tunnelprotocol.MaxStreamWindowBytes
	telemetryLogLevelInfo         = "info"
	telemetryLogLevelWarn         = "warn"
	telemetryLogLevelError        = "error"
	telemetryEventEgressStarted   = "egress_token_request_started"
	telemetryEventEgressCompleted = "egress_token_request_completed"
	telemetryEventEgressFailed    = "egress_token_request_failed"
	telemetryEventAgentSummary    = "agent_stream_summary"
	telemetryEventAgentExhausted  = "agent_stream_window_exhausted"
	telemetryEventAgentThreshold  = "agent_stream_window_threshold_crossed"
	telemetryEventControlDropped  = "bootstrap_control_message_dropped"
	telemetryEventFrameDropped    = "bootstrap_data_frame_dropped"
)

var agentStreamWindowThresholdBytes = []uint64{
	1024 * 1024,
	2 * 1024 * 1024,
	4 * 1024 * 1024,
	8 * 1024 * 1024,
	tunnelprotocol.AgentStreamWindowBytes * 95 / 100,
}

type telemetryRelayState struct {
	requested     bool
	opened        bool
	sendWindow    uint64
	bufferedLines [][]byte
	bufferedBytes uint64
}

func (session *LiveTunnelSession) openTelemetryStream(ctx context.Context) error {
	session.mutex.Lock()
	session.telemetry.requested = true
	session.telemetry.opened = false
	session.telemetry.sendWindow = 0
	session.mutex.Unlock()
	payload, err := telemetryOpenPayload()
	if err != nil {
		return err
	}
	return session.writeRawControl(ctx, payload)
}

func (session *LiveTunnelSession) closeTelemetryStream(ctx context.Context) error {
	session.mutex.Lock()
	wasRequested := session.telemetry.requested
	session.telemetry = telemetryRelayState{}
	session.mutex.Unlock()
	if !wasRequested {
		return nil
	}
	payload, err := telemetryClosePayload()
	if err != nil {
		return err
	}
	return session.writeRawControl(ctx, payload)
}

func (session *LiveTunnelSession) handleTelemetryControl(ctx context.Context, payload string) (bool, error) {
	message, err := parseTelemetryControl(payload)
	if err != nil {
		return false, nil
	}
	if message == nil {
		return false, nil
	}
	if message.streamID != SandboxTelemetryLogStreamID {
		return true, nil
	}
	switch message.messageType {
	case "telemetry.open.ok":
		session.mutex.Lock()
		session.telemetry.opened = true
		session.telemetry.sendWindow = message.initialWindowBytes
		session.mutex.Unlock()
		return true, session.flushTelemetry(ctx)
	case "telemetry.window":
		session.mutex.Lock()
		session.telemetry.sendWindow += message.bytes
		session.mutex.Unlock()
		return true, session.flushTelemetry(ctx)
	case "telemetry.open.error":
		session.mutex.Lock()
		session.telemetry = telemetryRelayState{}
		session.mutex.Unlock()
		return true, fmt.Errorf("gateway rejected telemetry stream: %s (%s)", message.message, message.code)
	case "telemetry.reset":
		session.mutex.Lock()
		session.telemetry = telemetryRelayState{}
		session.mutex.Unlock()
		return true, fmt.Errorf("gateway reset telemetry stream: %s (%s)", message.message, message.code)
	default:
		return true, nil
	}
}

func (session *LiveTunnelSession) recordTelemetryLog(ctx context.Context, level string, event string, extraFields map[string]any) error {
	line, err := encodeTelemetryLogLine(session.clock, level, event, extraFields)
	if err != nil {
		return err
	}
	return session.enqueueTelemetryLogLine(ctx, line)
}

func (session *LiveTunnelSession) recordBufferedTelemetryLog(level string, event string, extraFields map[string]any) error {
	line, err := encodeTelemetryLogLine(session.clock, level, event, extraFields)
	if err != nil {
		return err
	}
	return session.bufferTelemetryLogLine(line)
}

func (session *LiveTunnelSession) enqueueTelemetryLogLine(ctx context.Context, line string) error {
	if err := session.bufferTelemetryLogLine(line); err != nil {
		return err
	}
	return session.flushTelemetry(ctx)
}

func (session *LiveTunnelSession) bufferTelemetryLogLine(line string) error {
	session.mutex.Lock()
	if session.telemetry.bufferedBytes+uint64(len(line)) > telemetryBufferCapacityBytes {
		session.mutex.Unlock()
		return fmt.Errorf("telemetry buffer exceeded the configured capacity")
	}
	session.telemetry.bufferedLines = append(session.telemetry.bufferedLines, []byte(line))
	session.telemetry.bufferedBytes += uint64(len(line))
	session.mutex.Unlock()
	return nil
}

func (session *LiveTunnelSession) forwardSupervisorLifecycleEvents(ctx context.Context) {
	if session.supervisorHandle == nil {
		return
	}
	for _, line := range session.supervisorHandle.DrainForwardedLifecycleEventLines() {
		forwardedLine, err := supervision.EncodeForwardedLifecycleEventLogLine(line)
		if err != nil {
			fmt.Fprintf(os.Stderr, "sandboxd failed to encode forwarded lifecycle telemetry: %v\n", err)
			continue
		}
		if err := session.enqueueTelemetryLogLine(ctx, forwardedLine); err != nil {
			fmt.Fprintf(os.Stderr, "sandboxd failed to publish supervisor lifecycle telemetry: %v\n", err)
		}
	}
}

func (session *LiveTunnelSession) flushTelemetry(ctx context.Context) error {
	for {
		session.mutex.Lock()
		if !session.telemetry.opened || len(session.telemetry.bufferedLines) == 0 {
			session.mutex.Unlock()
			return nil
		}
		line := session.telemetry.bufferedLines[0]
		if session.telemetry.sendWindow < uint64(len(line)) {
			session.mutex.Unlock()
			return nil
		}
		session.telemetry.sendWindow -= uint64(len(line))
		session.telemetry.bufferedBytes -= uint64(len(line))
		session.telemetry.bufferedLines = session.telemetry.bufferedLines[1:]
		session.mutex.Unlock()
		encoded, err := tunnelprotocol.EncodeStreamDataFrame(SandboxTelemetryLogStreamID, tunnelprotocol.PayloadKindRawBytes, line)
		if err != nil {
			return err
		}
		session.writeMutex.Lock()
		err = session.connection.Write(ctx, websocket.MessageBinary, encoded)
		session.writeMutex.Unlock()
		if err != nil {
			return err
		}
	}
}

func (session *LiveTunnelSession) recordEgressTokenEvent(ctx context.Context, event string, requestID string, extraFields map[string]any) {
	fields := map[string]any{
		"component":         "TunnelSession",
		"sandboxInstanceId": session.sandboxInstanceID,
		"requestId":         requestID,
	}
	for key, value := range extraFields {
		fields[key] = value
	}
	level := telemetryLogLevelInfo
	if event == telemetryEventEgressFailed {
		level = telemetryLogLevelWarn
	}
	if err := session.recordTelemetryLog(ctx, level, event, fields); err != nil {
		fmt.Fprintf(os.Stderr, "sandboxd failed to publish telemetry event %q: %v\n", event, err)
	}
}

func (session *LiveTunnelSession) publishAgentStreamSummary(ctx context.Context, streamID uint32, stream *liveTunnelStream, outcome string, closeSource string, resetCode string, reason string) {
	if stream == nil || stream.agentStats == nil {
		return
	}
	fields := agentStreamSummaryFields(session.clock, streamID, stream, outcome, closeSource, resetCode, reason)
	if err := session.recordTelemetryLog(ctx, telemetryLogLevelInfo, telemetryEventAgentSummary, fields); err != nil {
		fmt.Fprintf(os.Stderr, "sandboxd failed to publish telemetry event %q: %v\n", telemetryEventAgentSummary, err)
	}
}

func (session *LiveTunnelSession) bufferAgentStreamSummary(streamID uint32, stream *liveTunnelStream, outcome string, closeSource string, resetCode string, reason string) {
	if stream == nil || stream.agentStats == nil {
		return
	}
	fields := agentStreamSummaryFields(session.clock, streamID, stream, outcome, closeSource, resetCode, reason)
	if err := session.recordBufferedTelemetryLog(telemetryLogLevelInfo, telemetryEventAgentSummary, fields); err != nil {
		fmt.Fprintf(os.Stderr, "sandboxd failed to buffer telemetry event %q: %v\n", telemetryEventAgentSummary, err)
	}
}

func agentStreamSummaryFields(clock timeutil.Clock, streamID uint32, stream *liveTunnelStream, outcome string, closeSource string, resetCode string, reason string) map[string]any {
	fields := map[string]any{
		"streamId":            uint64(streamID),
		"channelKind":         "agent",
		"outcome":             outcome,
		"closeSource":         closeSource,
		"durationMs":          clock.NowMS() - stream.agentStats.openedAtMS,
		"messageCountOut":     stream.agentStats.messageCountOut,
		"messageCountIn":      stream.agentStats.messageCountIn,
		"totalBytesOut":       stream.agentStats.totalBytesOut,
		"totalBytesIn":        stream.agentStats.totalBytesIn,
		"maxMessageBytesOut":  stream.agentStats.maxMessageBytesOut,
		"maxMessageBytesIn":   stream.agentStats.maxMessageBytesIn,
		"maxOutstandingBytes": stream.agentStats.maxOutstandingBytes,
		"avgCreditReturnMs":   stream.agentStats.avgCreditReturnMS(),
		"creditReturnCount":   stream.agentStats.creditReturnCount,
		"resetCode":           nil,
		"reason":              nil,
	}
	if resetCode != "" {
		fields["resetCode"] = resetCode
	}
	if reason != "" {
		fields["reason"] = reason
	}
	return fields
}

func (session *LiveTunnelSession) publishAgentWindowExhausted(ctx context.Context, streamID uint32, stream *liveTunnelStream, payloadKind byte, payloadBytes int) {
	if stream == nil || stream.agentStats == nil {
		return
	}
	fields := map[string]any{
		"streamId":                      uint64(streamID),
		"channelKind":                   "agent",
		"payloadKind":                   telemetryPayloadKindName(payloadKind),
		"payloadBytes":                  uint64(payloadBytes),
		"availableBytes":                stream.window,
		"outstandingBytes":              uint64(tunnelprotocol.AgentStreamWindowBytes) - stream.window,
		"maxWindowBytes":                uint64(tunnelprotocol.AgentStreamWindowBytes),
		"payloadExceedsMaxWindow":       uint64(payloadBytes) > uint64(tunnelprotocol.AgentStreamWindowBytes),
		"payloadExceedsAvailableWindow": uint64(payloadBytes) > stream.window,
		"messageCountOut":               stream.agentStats.messageCountOut,
		"streamAgeMs":                   session.clock.NowMS() - stream.agentStats.openedAtMS,
		"oldestUnackedMs":               stream.agentStats.oldestUnackedAgeMS(session.clock.NowMS()),
	}
	if err := session.recordTelemetryLog(ctx, telemetryLogLevelWarn, telemetryEventAgentExhausted, fields); err != nil {
		fmt.Fprintf(os.Stderr, "sandboxd failed to publish telemetry event %q: %v\n", telemetryEventAgentExhausted, err)
	}
}

func (session *LiveTunnelSession) publishAgentWindowThresholdCrossed(ctx context.Context, telemetry agentStreamThresholdTelemetry) {
	fields := map[string]any{
		"streamId":         uint64(telemetry.streamID),
		"channelKind":      "agent",
		"payloadKind":      telemetryPayloadKindName(telemetry.payloadKind),
		"payloadBytes":     uint64(telemetry.payloadBytes),
		"availableBytes":   telemetry.availableBytes,
		"outstandingBytes": telemetry.outstandingBytes,
		"thresholdBytes":   telemetry.thresholdBytes,
		"maxWindowBytes":   uint64(tunnelprotocol.AgentStreamWindowBytes),
		"messageCountOut":  telemetry.messageCountOut,
		"streamAgeMs":      telemetry.streamAgeMS,
		"oldestUnackedMs":  telemetry.oldestUnackedMS,
	}
	if err := session.recordTelemetryLog(ctx, telemetryLogLevelWarn, telemetryEventAgentThreshold, fields); err != nil {
		fmt.Fprintf(os.Stderr, "sandboxd failed to publish telemetry event %q: %v\n", telemetryEventAgentThreshold, err)
	}
}

func recordAgentInboundMessage(stream *liveTunnelStream, payloadBytes int) {
	if stream == nil || stream.agentStats == nil {
		return
	}
	bytes := uint64(payloadBytes)
	stream.agentStats.messageCountIn++
	stream.agentStats.totalBytesIn += bytes
	stream.agentStats.maxMessageBytesIn = max(stream.agentStats.maxMessageBytesIn, bytes)
}

func recordAgentOutboundMessage(stream *liveTunnelStream, payloadBytes int, sentAtMS uint64, outstandingBytes uint64) {
	if stream == nil || stream.agentStats == nil {
		return
	}
	bytes := uint64(payloadBytes)
	stream.agentStats.messageCountOut++
	stream.agentStats.totalBytesOut += bytes
	stream.agentStats.maxMessageBytesOut = max(stream.agentStats.maxMessageBytesOut, bytes)
	stream.agentStats.maxOutstandingBytes = max(stream.agentStats.maxOutstandingBytes, outstandingBytes)
	stream.agentStats.outstandingSends = append(stream.agentStats.outstandingSends, outstandingAgentSend{
		bytes:    bytes,
		sentAtMS: sentAtMS,
	})
}

func recordAgentCreditRestore(stream *liveTunnelStream, bytes uint64, restoredAtMS uint64) {
	if stream == nil || stream.agentStats == nil {
		return
	}
	remainingBytes := bytes
	for remainingBytes > 0 && len(stream.agentStats.outstandingSends) > 0 {
		frontSend := &stream.agentStats.outstandingSends[0]
		acknowledgedBytes := min(remainingBytes, frontSend.bytes)
		frontSend.bytes -= acknowledgedBytes
		remainingBytes -= acknowledgedBytes
		stream.agentStats.creditReturnCount++
		stream.agentStats.creditReturnTotalMS += restoredAtMS - min(restoredAtMS, frontSend.sentAtMS)
		if frontSend.bytes == 0 {
			stream.agentStats.outstandingSends = stream.agentStats.outstandingSends[1:]
		}
	}
}

func (stats *agentStreamStats) avgCreditReturnMS() any {
	if stats.creditReturnCount == 0 {
		return nil
	}
	return stats.creditReturnTotalMS / stats.creditReturnCount
}

func (stats *agentStreamStats) oldestUnackedAgeMS(nowMS uint64) any {
	if len(stats.outstandingSends) == 0 {
		return nil
	}
	sentAtMS := stats.outstandingSends[0].sentAtMS
	return nowMS - min(nowMS, sentAtMS)
}

func (stats *agentStreamStats) takeNewThresholdCrossings(outstandingBytes uint64) []uint64 {
	var crossed []uint64
	for index, thresholdBytes := range agentStreamWindowThresholdBytes {
		emissionBit := uint8(1 << index)
		if outstandingBytes >= thresholdBytes && stats.thresholdMask&emissionBit == 0 {
			stats.thresholdMask |= emissionBit
			crossed = append(crossed, thresholdBytes)
		}
	}
	return crossed
}

func telemetryPayloadKindName(payloadKind byte) string {
	switch payloadKind {
	case tunnelprotocol.PayloadKindWebSocketText:
		return "websocket_text"
	case tunnelprotocol.PayloadKindWebSocketBinary:
		return "websocket_binary"
	case tunnelprotocol.PayloadKindRawBytes:
		return "raw_bytes"
	default:
		return "unknown"
	}
}

func encodeTelemetryLogLine(clock timeutil.Clock, level string, event string, extraFields map[string]any) (string, error) {
	if clock == nil {
		return "", fmt.Errorf("telemetry log clock is required")
	}
	switch level {
	case telemetryLogLevelInfo, telemetryLogLevelWarn, telemetryLogLevelError:
	default:
		return "", fmt.Errorf("sandbox telemetry log level %q is unsupported", level)
	}
	if event == "" {
		return "", fmt.Errorf("sandbox telemetry log event must not be empty")
	}
	payload := map[string]any{
		"timestamp": timeutil.FormatRFC3339Timestamp(clock.NowSystemTime()),
		"level":     level,
		"event":     event,
	}
	for key, value := range extraFields {
		if key == "" {
			return "", fmt.Errorf("sandbox telemetry log field name must not be empty")
		}
		switch key {
		case "timestamp", "level", "event":
			return "", fmt.Errorf("sandbox telemetry log field %q is reserved", key)
		}
		if !telemetryScalar(value) {
			return "", fmt.Errorf("sandbox telemetry log field %q must be scalar", key)
		}
		payload[key] = value
	}
	serialized, err := json.Marshal(payload)
	if err != nil {
		return "", err
	}
	return string(serialized) + "\n", nil
}

func telemetryScalar(value any) bool {
	switch value.(type) {
	case nil, bool, string, float64, float32, int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64, json.Number:
		return true
	default:
		return false
	}
}

type telemetryControlMessage struct {
	messageType        string
	streamID           uint32
	initialWindowBytes uint64
	bytes              uint64
	code               string
	message            string
}

func parseTelemetryControl(payload string) (*telemetryControlMessage, error) {
	var raw struct {
		Type               string `json:"type"`
		StreamID           uint32 `json:"streamId"`
		InitialWindowBytes uint64 `json:"initialWindowBytes"`
		Bytes              uint64 `json:"bytes"`
		Code               string `json:"code"`
		Message            string `json:"message"`
	}
	if err := json.Unmarshal([]byte(payload), &raw); err != nil {
		return nil, err
	}
	switch raw.Type {
	case "telemetry.open.ok":
		if raw.InitialWindowBytes == 0 {
			return nil, fmt.Errorf("telemetry.open.ok initialWindowBytes must be a positive integer")
		}
	case "telemetry.window":
		if raw.Bytes == 0 {
			return nil, fmt.Errorf("telemetry.window bytes must be a positive integer")
		}
	case "telemetry.open.error", "telemetry.reset":
		if raw.Code == "" || raw.Message == "" {
			return nil, fmt.Errorf("%s code and message are required", raw.Type)
		}
	default:
		if len(raw.Type) >= len("telemetry.") && raw.Type[:len("telemetry.")] == "telemetry." {
			return &telemetryControlMessage{messageType: raw.Type, streamID: raw.StreamID}, nil
		}
		return nil, nil
	}
	return &telemetryControlMessage{
		messageType:        raw.Type,
		streamID:           raw.StreamID,
		initialWindowBytes: raw.InitialWindowBytes,
		bytes:              raw.Bytes,
		code:               raw.Code,
		message:            raw.Message,
	}, nil
}

func telemetryOpenPayload() (string, error) {
	return marshalTelemetryControlPayload(map[string]any{
		"type":     "telemetry.open",
		"streamId": SandboxTelemetryLogStreamID,
		"signal":   telemetryLogsSignal,
		"format":   telemetryLogsFormat,
	})
}

func telemetryClosePayload() (string, error) {
	return marshalTelemetryControlPayload(map[string]any{
		"type":     "telemetry.close",
		"streamId": SandboxTelemetryLogStreamID,
	})
}

func marshalTelemetryControlPayload(payload map[string]any) (string, error) {
	serialized, err := json.Marshal(payload)
	if err != nil {
		return "", err
	}
	return string(serialized), nil
}
